README.md
[![Latest Stable Version](http://poser.pugx.org/nguyenanhung/slug-helper/v)](https://packagist.org/packages/nguyenanhung/slug-helper) [![Total Downloads](http://poser.pugx.org/nguyenanhung/slug-helper/downloads)](https://packagist.org/packages/nguyenanhung/slug-helper) [![Latest Unstable Version](http://poser.pugx.org/nguyenanhung/slug-helper/v/unstable)](https://packagist.org/packages/nguyenanhung/slug-helper) [![License](http://poser.pugx.org/nguyenanhung/slug-helper/license)](https://packagist.org/packages/nguyenanhung/slug-helper) [![PHP Version Require](http://poser.pugx.org/nguyenanhung/slug-helper/require/php)](https://packagist.org/packages/nguyenanhung/slug-helper)

# Slug Helper

Slug Helper - Basic, Simple and Lightweight

## Version

- [x] `v1.x` support all PHP version `>=5.5.9`
- [x] `v2.x` support all PHP version `>=7.0`

## Contact & Support

If any question & request, please contact following information

| Name        | Email                | Skype            | Facebook      |
| ----------- | -------------------- | ---------------- | ------------- |
| Hung Nguyen | dev@nguyenanhung.com | nguyenanhung5891 | @nguyenanhung |

From Vietnam with Love <3


CHANGELOG.md
# CHANGELOG

## [1.0.0] - 2021/09/25

### Added

- [x] Init Project

## Contact & Support

If any question & request, please contact following information

| Name        | Email                | Skype            | Facebook      |
| ----------- | -------------------- | ---------------- | ------------- |
| Hung Nguyen | dev@nguyenanhung.com | nguyenanhung5891 | @nguyenanhung |


test/test.php
<?php
require_once __DIR__ . '/../vendor/autoload.php';

use nguyenanhung\Libraries\Slug\SlugUrl;

$slugUrl = new SlugUrl();
$slugUrl->setSiteUrl('https://nguyenanhung.com')->setSiteExt('.html');

$name = 'Nguyễn An Hưng';

echo "slugify: " . $slugUrl->slugify($name) . PHP_EOL;
echo "searchSlugify: " . $slugUrl->searchSlugify($name) . PHP_EOL;
echo "strToEn: " . $slugUrl->strToEn($name) . PHP_EOL;


helpers/helpers.php
<?php

use nguyenanhung\Libraries\Slug\SlugUrl;

if (!function_exists('seoToSlugify')) {
    /**
     * Function seoToSlugify
     *
     * @param $str
     *
     * @return string
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 09/25/2021 34:34
     */
    function seoToSlugify($str): string
    {
        return (new SlugUrl())->slugify($str);
    }
}
if (!function_exists('seoSearchSlugify')) {
    /**
     * Function seoSearchSlugify
     *
     * @param $str
     *
     * @return string
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 09/25/2021 34:34
     */
    function seoSearchSlugify($str): string
    {
        return (new SlugUrl())->searchSlugify($str);
    }
}
if (!function_exists('convertStrToEn')) {
    /**
     * Function convertStrToEn
     *
     * @param $str
     *
     * @return string
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 09/25/2021 34:34
     */
    function convertStrToEn($str): string
    {
        return (new SlugUrl())->toEnglish($str);
    }
}


src/Repository/DataRepository.php
<?php
/**
 * Project slug-helper
 * Created by PhpStorm
 * User: 713uk13m <dev@nguyenanhung.com>
 * Copyright: 713uk13m <dev@nguyenanhung.com>
 * Date: 09/25/2021
 * Time: 08:24
 */

namespace nguyenanhung\Libraries\Slug\Repository;

/**
 * Class DataRepository
 *
 * @package   nguyenanhung\Libraries\Slug\Repository
 * @author    713uk13m <dev@nguyenanhung.com>
 * @copyright 713uk13m <dev@nguyenanhung.com>
 */
class DataRepository
{

    const CONFIG_PATH = 'config';
    const CONFIG_EXT  = '.php';

    /**
     * Hàm lấy nội dung config được quy định trong thư mục config
     *
     * @author: 713uk13m <dev@nguyenanhung.com>
     * @time  : 9/28/18 14:47
     *
     * @param string $configName Tên file config
     *
     * @return array|mixed
     */
    public static function getData($configName)
    {
        $path = __DIR__ . DIRECTORY_SEPARATOR . self::CONFIG_PATH . DIRECTORY_SEPARATOR . $configName . self::CONFIG_EXT;
        if (is_file($path) && file_exists($path)) {
            return require $path;
        }

        return array();
    }

    /**
     * Hàm lấy nội dung Data từ 1 file bất kỳ trong hệ thống
     *
     * @author: 713uk13m <dev@nguyenanhung.com>
     * @time  : 10/17/18 09:25
     *
     * @param string $filename Đường dẫn file config
     *
     * @return array|mixed
     */
    public static function getDataContent($filename)
    {
        if (is_file($filename) && file_exists($filename)) {
            return require $filename;
        }

        return array();
    }
}


src/SlugUrl.php
<?php
/**
 * Project slug-helper
 * Created by PhpStorm
 * User: 713uk13m <dev@nguyenanhung.com>
 * Copyright: 713uk13m <dev@nguyenanhung.com>
 * Date: 09/25/2021
 * Time: 08:24
 */

namespace nguyenanhung\Libraries\Slug;

use Exception;
use Cocur\Slugify\Slugify;
use nguyenanhung\Libraries\Slug\Repository\DataRepository;

/**
 * Class SlugUrl
 *
 * @package   nguyenanhung\Libraries\Slug
 * @author    713uk13m <dev@nguyenanhung.com>
 * @copyright 713uk13m <dev@nguyenanhung.com>
 */
class SlugUrl
{
    /** @var string $siteUrl */
    protected $siteUrl;

    /** @var string Site Ext */
    private $siteExt = '.html';

    /**
     * Function setSiteUrl
     *
     * @param string $siteUrl
     *
     * @return $this
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 09/25/2021 26:03
     */
    public function setSiteUrl(string $siteUrl = ''): SlugUrl
    {
        $this->siteUrl = $siteUrl;

        return $this;
    }

    /**
     * Function getSiteUrl
     *
     * @return string
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 09/25/2021 26:11
     */
    public function getSiteUrl(): string
    {
        return $this->siteUrl;
    }

    /**
     * Function setSiteExt
     *
     * @param $siteExt
     *
     * @return $this
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 09/25/2021 26:00
     */
    public function setSiteExt($siteExt): SlugUrl
    {
        $this->siteExt = $siteExt;

        return $this;
    }

    /**
     * Function getSiteExt
     *
     * @return string
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 09/25/2021 26:17
     */
    public function getSiteExt(): string
    {
        return $this->siteExt;
    }

    /**
     * Function slugify - SEO Slugify
     *
     * @author: 713uk13m <dev@nguyenanhung.com>
     * @time  : 9/21/18 02:49
     *
     * @param string $str
     *
     * @return string
     */
    public function slugify(string $str = ''): string
    {
        try {
            $slugify = new Slugify();

            return $slugify->slugify($str);
        } catch (Exception $e) {
            return $this->convertVietnameseToEnglish($str);
        }

    }

    /**
     * Function searchSlugify - SEO Search Slugify
     *
     * @param string $str
     *
     * @return string
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 09/25/2021 30:20
     */
    public function searchSlugify(string $str = ''): string
    {
        try {
            $options = array('separator' => '+');
            $slugify = new Slugify($options);

            return $slugify->slugify($str);
        } catch (Exception $e) {
            return $this->convertVietnameseToEnglish($str);
        }
    }

    /**
     * Function toEnglish - Convert String to English
     *
     * @param string $str
     *
     * @return string
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 09/25/2021 32:49
     */
    public function toEnglish(string $str = ''): string
    {
        try {
            $options = array('separator' => ' ');
            $slugify = new Slugify($options);

            return $slugify->slugify($str);
        } catch (Exception $e) {
            return $this->convertVietnameseToEnglish($str);
        }
    }

    /**
     * Function urlEncode
     *
     * @param string $url
     *
     * @return string
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 09/25/2021 31:43
     */
    public function urlEncode(string $url = ''): string
    {
        return urlencode($url);
    }

    /**
     * Function urlDecode
     *
     * @param string $url
     *
     * @return string
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 09/25/2021 31:46
     */
    public function urlDecode(string $url): string
    {
        return urldecode($url);
    }

    /**
     * Function convertVietnameseToEnglish
     *
     * Hàm chuyển đổi ký tự từ tiếng Việt, và các ký tự đặc biệt sang ký tự không dấu
     *
     * Sử dụng trong trường hợp class slugify nó không chạy
     *
     * @param string $str Chuỗi ký tự đầu vào
     *
     * @return string Đầu ra rà 1 chuỗi ký tự
     * @author: 713uk13m <dev@nguyenanhung.com>
     * @time  : 10/13/18 01:17
     *
     */
    public function convertVietnameseToEnglish(string $str = ''): string
    {
        $str  = trim($str);
        $str  = function_exists('mb_strtolower') ? mb_strtolower($str) : strtolower($str);
        $data = DataRepository::getData('convert_vi_to_en');
        if (!empty($str)) {
            $str = str_replace($data['vn_array'], $data['en_array'], $str);
            $str = str_replace($data['special_array'], $data['separator'], $str);
            $str = str_replace(' ', $data['separator'], $str);
            while (strpos($str, '--') > 0) {
                $str = str_replace('--', $data['separator'], $str);
            }
            while (strpos($str, '--') === 0) {
                $str = str_replace('--', $data['separator'], $str);
            }
        }

        return $str;
    }
}
